## Pearson Product Moment Correlation Coefficient


A correlation is the degree of association between two variables. One of my favorite books on this topic is: 

<img src="http://zacharski.org/files/courses/cs419/mg_statistics_big.png" width="250" />

and they illustrate it by looking at
## Ladies expenditures on clothes and makeup

<img src="http://zacharski.org/files/courses/cs419/correlation1.png" width="650" />
<img src="http://zacharski.org/files/courses/cs419/corr2.png" width="650" />

So let's go ahead and create that data in pandas and show the table:

In [1]:
import pandas as pd
from pandas import DataFrame
makeup =  [3000, 5000, 12000, 2000, 7000,  15000,  5000,  6000, 8000,  10000]
clothes = [7000, 8000, 25000, 5000, 12000, 30000, 10000, 15000, 20000, 18000]
ladies = ['Ms A','Ms B','Ms C','Ms D','Ms E','Ms F','Ms G','Ms H','Ms I','Ms J',]
monthly = DataFrame({'makeup': makeup, 'clothes': clothes}, index= ladies)
monthly

makeup  clothes
Ms A    3000     7000
Ms B    5000     8000
Ms C   12000    25000
Ms D    2000     5000
Ms E    7000    12000
Ms F   15000    30000
Ms G    5000    10000
Ms H    6000    15000
Ms I    8000    20000
Ms J   10000    18000

and let's show the scatterplot

In [2]:
from bokeh.plotting import figure, output_file, show
from bokeh.io import push_notebook, show, output_notebook
output_notebook()
x = figure(title="Montly Expenditures on Makeup and Clothes", x_axis_label="Money spent on makeup", y_axis_label="Money spent on clothes")
x.circle(monthly['makeup'], monthly['clothes'], size=10, color="navy", alpha=0.5)
output_file("stuff.html")

show(x)

Loading BokehJS ...

When the data points are close to a straight line going up, we say that there is a positive correlation between the two variables. Let's look at a few more examples:

## Weight and calories in 1-3 yr/old children


In [3]:
weight = [7.7, 7.8, 8.6, 8.5, 8.6, 9, 10.1, 11.5, 11, 10.2, 11.9, 10.4, 9.3, 9.1, 8.5, 11]
calories = [360, 400, 500, 370, 525, 800, 900, 1200, 1000, 1400, 1600, 850, 575, 425, 950, 800]
kids = DataFrame({'weight': weight, 'calories': calories})
kids

weight  calories
0      7.7       360
1      7.8       400
2      8.6       500
3      8.5       370
4      8.6       525
5      9.0       800
6     10.1       900
7     11.5      1200
8     11.0      1000
9     10.2      1400
10    11.9      1600
11    10.4       850
12     9.3       575
13     9.1       425
14     8.5       950
15    11.0       800

In [4]:
p = figure(title="Weight and calories in 1-3 yr.old children", 
           x_axis_label="weight (kg)", y_axis_label='weekly calories')
p.circle(kids['weight'], kids['calories'], size=10, color='navy', alpha=0.5)
show(p)


## The stronger the correlation the closer to a straight line:


In [5]:
midterm = [65, 70, 72, 75, 78, 78, 80, 83, 87, 87, 90, 90, 92, 93, 95, 100]
final =   [72, 68, 75, 76, 79, 73, 79, 85, 89, 87, 92, 93, 94, 91, 97, 98]
grades = DataFrame({'midterm': midterm, 'final': final})
grades

midterm  final
0        65     72
1        70     68
2        72     75
3        75     76
4        78     79
5        78     73
6        80     79
7        83     85
8        87     89
9        87     87
10       90     92
11       90     93
12       92     94
13       93     91
14       95     97
15      100     98

In [6]:
p = figure(title="Correlation of Midterm to Final Grades", 
           x_axis_label="Midterm Grade", y_axis_label='Final Grade')
p.circle(grades['midterm'], grades['final'], size=10, color='navy', alpha=0.5)
show(p)

## perfect positive correlation

Here I am comparing the heights in inches of some NHL players with their heights in cm.

In [7]:
inches =[68, 73, 69,72,71,77]
cm = [173, 185, 175, 183, 180, 196]
nhlHeights = DataFrame({'heightInches': inches, 'heightCM': cm})
nhlHeights

heightInches  heightCM
0            68       173
1            73       185
2            69       175
3            72       183
4            71       180
5            77       196

In [8]:
p = figure(title="Comparison of Height in Inches and Height in CM", 
           x_axis_label="Height in Inches", 
           y_axis_label="Height in centimeters")
p.circle(nhlHeights['heightInches'], nhlHeights['heightCM'],
         size=10, color='navy', alpha=0.5)
show(p)

## No correlation = far from straight line
Comparing length of hometown name to height in inches.

In [9]:
medicineHat = pd.read_csv('https://raw.githubusercontent.com/zacharski/machine-learning/master/data/medicineHatTigers.csv')
medicineHat

Name Pos. L/R  Height  Weight  Birthday          Hometown
0       Brennan Bosch    C   R      68     173  02/14/88  Martensville, SK
1        Scott Wasden    C   R      73     188  01/04/88      Westbank, BC
2        Colton Grant   LW   L      69     177  03/20/89      Standard, AB
3         Darren Helm   LW   L      72     182  01/21/87   St. Andrews, MB
4       Derek Dorsett   RW   L      71     178  12/20/86    Kindersley, SK
5          Daine Todd    C   R      70     173  01/10/87      Red Deer, AB
6       Tyler Swystun   RW   R      71     185  01/15/88      Cochrane, AB
7          Matt Lowry    C   R      72     186  03/02/88       Neepawa, MB
8   Kevin Undershute    LW   L      72     178  04/12/87  Medicine Hat, AB
9        Jerrid Sauer   RW   R      71     196  09/12/87  Medicine Hat, AB
10        Tyler Ennis    C   L      69     160  10/06/89      Edmonton, AB
11    Jordan Hickmott    C   R      72     183  04/11/90        Mission BC
12       Jakub Rumpel   RW   R      68     166  01/27/87  Hrnciarovce, SLO
13    Bretton Cameron    C   R      71     168  01/26/89      Didsbury, AB
14      Chris Stevens   LW   L      70     197  08/20/86  Dawson Creek, BC
15       Gord Baldwin    D   L      77     205  03/01/87      Winnipeg, MB
16     David Schlemko    D   L      73     195  05/07/87      Edmonton, AB
17       Trever Glass    D   L      72     190  01/22/88      Cochrane, AB
18       Kris Russell    D   L      70     177  05/02/87      Caroline, AB
19     Michael Sauer     D   R      73     205  08/07/87       Sartell, MN
20     Mark Isherwood    D   R      72     183  01/31/89    Abbotsford, BC
21      Shayne Brown     D   L      73     198  02/20/89   Stony Plain, AB
22    Jordan Bendfeld    D   R      75     230  02/09/88         Ledue, AB
23       Ryan Holfeld    G   L      71     166  06/29/89         LeRoy, SK
24       Matt Keetley    G   R      74     189  04/27/86  Medicine Hat, AB

In [10]:
medicineHat['hometownLength'] = medicineHat['Hometown'].str.len()
medicineHat

Name Pos. L/R  Height  Weight  Birthday          Hometown  \
0       Brennan Bosch    C   R      68     173  02/14/88  Martensville, SK   
1        Scott Wasden    C   R      73     188  01/04/88      Westbank, BC   
2        Colton Grant   LW   L      69     177  03/20/89      Standard, AB   
3         Darren Helm   LW   L      72     182  01/21/87   St. Andrews, MB   
4       Derek Dorsett   RW   L      71     178  12/20/86    Kindersley, SK   
5          Daine Todd    C   R      70     173  01/10/87      Red Deer, AB   
6       Tyler Swystun   RW   R      71     185  01/15/88      Cochrane, AB   
7          Matt Lowry    C   R      72     186  03/02/88       Neepawa, MB   
8   Kevin Undershute    LW   L      72     178  04/12/87  Medicine Hat, AB   
9        Jerrid Sauer   RW   R      71     196  09/12/87  Medicine Hat, AB   
10        Tyler Ennis    C   L      69     160  10/06/89      Edmonton, AB   
11    Jordan Hickmott    C   R      72     183  04/11/90        Mission BC   
12       Jakub Rumpel   RW   R      68     166  01/27/87  Hrnciarovce, SLO   
13    Bretton Cameron    C   R      71     168  01/26/89      Didsbury, AB   
14      Chris Stevens   LW   L      70     197  08/20/86  Dawson Creek, BC   
15       Gord Baldwin    D   L      77     205  03/01/87      Winnipeg, MB   
16     David Schlemko    D   L      73     195  05/07/87      Edmonton, AB   
17       Trever Glass    D   L      72     190  01/22/88      Cochrane, AB   
18       Kris Russell    D   L      70     177  05/02/87      Caroline, AB   
19     Michael Sauer     D   R      73     205  08/07/87       Sartell, MN   
20     Mark Isherwood    D   R      72     183  01/31/89    Abbotsford, BC   
21      Shayne Brown     D   L      73     198  02/20/89   Stony Plain, AB   
22    Jordan Bendfeld    D   R      75     230  02/09/88         Ledue, AB   
23       Ryan Holfeld    G   L      71     166  06/29/89         LeRoy, SK   
24       Matt Keetley    G   R      74     189  04/27/86  Medicine Hat, AB   

    hometownLength  
0               16  
1               12  
2               12  
3               15  
4               14  
5               12  
6               12  
7               11  
8               16  
9               16  
10              12  
11              10  
12              16  
13              12  
14              16  
15              12  
16              12  
17              12  
18              12  
19              11  
20              14  
21              15  
22               9  
23               9  
24              16

In [11]:
medicineHat[['Height','hometownLength']]

Height  hometownLength
0       68              16
1       73              12
2       69              12
3       72              15
4       71              14
5       70              12
6       71              12
7       72              11
8       72              16
9       71              16
10      69              12
11      72              10
12      68              16
13      71              12
14      70              16
15      77              12
16      73              12
17      72              12
18      70              12
19      73              11
20      72              14
21      73              15
22      75               9
23      71               9
24      74              16

In [12]:
p = figure(title="Correlation of the number of Letters in the Hometown to Height", 
           x_axis_label="Player's Height", y_axis_label="Hometown Name Length")
p.circle(medicineHat['Height'], medicineHat['hometownLength'], size=10, color='navy', alpha=0.5)
show(p)

# You graph the next 2

## negative correlation has a line going downhill

#### the relationship of hair length to a person's height

In [13]:
height =[62, 64, 65, 68, 69, 70, 67, 65, 72, 73, 74]
hairLength = [7, 10, 6, 4, 5, 4, 5, 8, 1, 1, 3]
cm = [173, 185, 175, 183, 180, 196]
people = DataFrame({'height': height, 'hairLength': hairLength})

#TBD

The taller someone is, the shorter their hair.


##  The Effects of Alcohol on the Dexterity Score
#### a strong negative correlation

In [14]:
drinks = list(range(11))
dexterity = [96, 96, 90, 63, 65, 50, 47, 46, 18, 17, 9]
drinking = DataFrame({'drinks': drinks, 'dexterity': dexterity})


# Numeric Represenstation of the Strength of the Correlation


## Pearson Correlation Coefficient

This ranges from -1 to 1.
1 is perfect positive correlation, -1 is perfect negative.

$$r=\frac{\sum_{i=1}^n(x_i - \bar{x})(y_i-\bar{y})}{\sqrt{\sum_{i=1}^n(x_i - \bar{x})}  \sqrt{\sum_{i=1}^n(y_i - \bar{y})}}$$

You should know how to compute the formula by hand. For help, checkout the textbook. In Pandas it is very easy to compute.

### Japanese ladies expensives on makeup and clothes
Let's go back to our first example.
First here is the data:

In [15]:
monthly

makeup  clothes
Ms A    3000     7000
Ms B    5000     8000
Ms C   12000    25000
Ms D    2000     5000
Ms E    7000    12000
Ms F   15000    30000
Ms G    5000    10000
Ms H    6000    15000
Ms I    8000    20000
Ms J   10000    18000

And our Scatter plot looked like this:

In [16]:

p = figure(title="Montly Expenditures on Makeup and Clothes", 
           x_axis_label="Money spent on makeup", y_axis_label="Money spent on clothes")
p.circle(monthly['makeup'], monthly['clothes'], size=10, color='navy', alpha=0.5)
show(p)


So that looks like a pretty strong positive correlation. To compute Pearson on this data we do:

In [17]:
monthly.corr()

makeup  clothes
makeup   1.00000  0.96802
clothes  0.96802  1.00000

There is the obvious: clothes expenditures are perfectly correlated with clothes expenditures. But the correlation between clothes and makeup is also very strong at 0.96802.

What about my idea that the number of characters in the name of someone's hometown is correlated with the height of that person:


In [18]:
medicineHat.corr()

Height    Weight  hometownLength
Height          1.000000  0.736502       -0.277811
Weight          0.736502  1.000000       -0.156846
hometownLength -0.277811 -0.156846        1.000000

The Pearson Correlation Coefficient for height to hometownLength is -0.2778 and that is pretty weak.

Finally, what about the number of drinks to dexterity:

In [19]:
drinking.corr()

drinks  dexterity
drinks     1.000000  -0.979237
dexterity -0.979237   1.000000

So -0.97923 indicates a very strong negative correlation

<h2 style="color:red">You Try</h2>
<h3 style="color:red">1. Medicine Hat Tigers 10xp</h3>
<span style="color:red">We are interested in whether there is a relationship between the height and weight of the Medicine Hat Players.</span>

<h4 style="color:red">a. first show the scatter plot</h4>


In [20]:
# your code here

<h4 style="color:red">b. Now compute Pearson</h4>

In [21]:
# your code here


So what do you think? Is there a correlation between height and weight?

<h3 style="color:red">2. A School Music Program: A case study  15xp</h3>
<span style="color:red">We are interested in seeing whether a relationship exists between music performance
grades and students' individual Iowa Test of Basic Skills (ITBS) scores. Using the
 test set at</span>



In [22]:
musicFile = "https://raw.githubusercontent.com/zacharski/machine-learning/master/data/musicITBS.csv"

<span style="color:red">Calculate (using Pearson) the strength of the association between music and language,
music and science, and music and math. Are any of these associations strong?</span>



In [23]:
# Your code here


<span style="color:red">Please graph (using a scatter chart) the strongest association</span>



<h3 style="color:red">3. Happiness Data 75xp</h3>

In the Happiness lab we investigated the [2017 edition Better Life Index](https://stats.oecd.org/index.aspx?DataSetCode=BLI#).  And examined if a subset of attributes could predict Life satisfaction

Now we are interested in examining all the columns of the data and finding which ones are most highly correlated with life satisfaction.

<span style="color:red">Here is what we need: 1)We would like to see the plots showing the relationship of each of the feature columns (student skills, education attainment, ...) to Life satisfaction. 2) We would also like to see the correlation values. 3) Which attributes are most correlated to life satisfaction? </span>

## Helpful comments

1. df.corr()  returns a pandas dataframe (this is useful to know if you want to just extract the correlations of one column to all the others)
2. Sorting may be helpful in this task. Here is an example of sorting the Hockey players by their height with tallest first.

In [25]:
medicineHat.sort_values(by=['Weight'], ascending=False)


Name Pos. L/R  Height  Weight  Birthday          Hometown  \
22    Jordan Bendfeld    D   R      75     230  02/09/88         Ledue, AB   
19     Michael Sauer     D   R      73     205  08/07/87       Sartell, MN   
15       Gord Baldwin    D   L      77     205  03/01/87      Winnipeg, MB   
21      Shayne Brown     D   L      73     198  02/20/89   Stony Plain, AB   
14      Chris Stevens   LW   L      70     197  08/20/86  Dawson Creek, BC   
9        Jerrid Sauer   RW   R      71     196  09/12/87  Medicine Hat, AB   
16     David Schlemko    D   L      73     195  05/07/87      Edmonton, AB   
17       Trever Glass    D   L      72     190  01/22/88      Cochrane, AB   
24       Matt Keetley    G   R      74     189  04/27/86  Medicine Hat, AB   
1        Scott Wasden    C   R      73     188  01/04/88      Westbank, BC   
7          Matt Lowry    C   R      72     186  03/02/88       Neepawa, MB   
6       Tyler Swystun   RW   R      71     185  01/15/88      Cochrane, AB   
11    Jordan Hickmott    C   R      72     183  04/11/90        Mission BC   
20     Mark Isherwood    D   R      72     183  01/31/89    Abbotsford, BC   
3         Darren Helm   LW   L      72     182  01/21/87   St. Andrews, MB   
8   Kevin Undershute    LW   L      72     178  04/12/87  Medicine Hat, AB   
4       Derek Dorsett   RW   L      71     178  12/20/86    Kindersley, SK   
2        Colton Grant   LW   L      69     177  03/20/89      Standard, AB   
18       Kris Russell    D   L      70     177  05/02/87      Caroline, AB   
0       Brennan Bosch    C   R      68     173  02/14/88  Martensville, SK   
5          Daine Todd    C   R      70     173  01/10/87      Red Deer, AB   
13    Bretton Cameron    C   R      71     168  01/26/89      Didsbury, AB   
23       Ryan Holfeld    G   L      71     166  06/29/89         LeRoy, SK   
12       Jakub Rumpel   RW   R      68     166  01/27/87  Hrnciarovce, SLO   
10        Tyler Ennis    C   L      69     160  10/06/89      Edmonton, AB   

    hometownLength  
22               9  
19              11  
15              12  
21              15  
14              16  
9               16  
16              12  
17              12  
24              16  
1               12  
7               11  
6               12  
11              10  
20              14  
3               15  
8               16  
4               14  
2               12  
18              12  
0               16  
5               12  
13              12  
23               9  
12              16  
10              12